### ISSUE TITLE
Looking for an example with custom metric for early_stop

### ISSUE THREAD

**Issue by jiwidi** [PyTorchLightning/pytorch-lightning#3505]
#### What is your question?
Hi!

I was checking the docs and `early_stop_callback` and the only example I could find was https://pytorch-lightning.readthedocs.io/en/0.4.9/examples/Examples/#trainer-example. Is there any sample code where this is used? Im looking for a Lightning model that returns a custom metric and the early stop is based on that metric.

For my case, im working on NLP and the metric WER(Word error rate) is my main focus during training. I would like to configure the training so it stops when WER in the test set doesnt decrease for 2 epochs. Is this possible?

### Code

This is my test step if it helps, right now I'm logging WER but my training loop doesnt use it for early stop
```python
def test_step(self, batch, batch_idx):
        spectrograms, labels, input_lengths, label_lengths = batch
        y_hat = self(spectrograms)  # (batch, time, n_class)
        output = F.log_softmax(y_hat, dim=2)
        output = output.transpose(0, 1)  # (time, batch, n_class)

        loss = self.criterion(output, labels, input_lengths, label_lengths)

        decoded_preds, decoded_targets = GreedyDecoder(output.transpose(0, 1), labels, label_lengths)
        n_correct_pred = sum([int(a == b) for a, b in zip(decoded_preds, decoded_targets)])

        test_cer, test_wer = [], []
        for j in range(len(decoded_preds)):
            test_cer.append(cer(decoded_targets[j], decoded_preds[j]))
            test_wer.append(wer(decoded_targets[j], decoded_preds[j]))
        avg_cer = sum(test_cer) / len(test_cer)
        avg_wer = sum(test_wer) / len(test_wer)
        logs = {
            "cer": avg_cer,
            "wer": avg_wer,
        }
        return {
            "val_loss": loss,
            "n_correct_pred": n_correct_pred,
            "n_pred": len(spectrograms),
            "log": logs,
        }
```
#### What's your environment?

 - OS: [Linux, ]
 - Packaging [pip, ]
 - Version [0.9.0]

**Comment by carmocca** [comment]
Hello Jaime! Haven't seen you in a long time, hope everything is ok! :smile: 

I have the exact example you are looking for. You can have a look here:
https://github.com/carmocca/PyLaia/blob/33455b4efaeebfee702fd384ef8f0c5ea580737d/laia/engine/htr_engine_module.py#L97-L107

We have implemented returning a `Result` object, please check out the docs [here](https://pytorch-lightning.readthedocs.io/en/latest/results.html)

**Comment by jiwidi** [comment]
Hello @carmocca! 

Yes it's been a long time, hope everything goes good with you too! Is nice to see you around here, means you are doing cool stuff! 

The example and docs solved my question, perfect! :+1:
